linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -0,0 +1,2206 @@
+      // fast path
+      fb += LocalGet(typeDataParam)
+      fb += StructGet(genTypeID.typeData, genFieldID.typeData.classOfValue)
+      fb += BrOnNonNull(alreadyInitializedLabel)
+      // slow path
+      fb += LocalGet(typeDataParam)
+      fb += Call(genFunctionID.createClassOf)
+    } // end bock alreadyInitializedLabel
+
+    fb.buildAndAddToModule()
+  }
+
+  /** `arrayTypeData: (ref typeData), i32 -> (ref vtable.java.lang.Object)`.
+   *
+   *  Returns the typeData/vtable of an array with `dims` dimensions over the given typeData. `dims`
+   *  must be be strictly positive.
+   */
+  private def genArrayTypeData()(implicit ctx: WasmContext): Unit = {
+    val typeDataType = RefType(genTypeID.typeData)
+    val objectVTableType = RefType(genTypeID.ObjectVTable)
+
+    /* Array classes extend Cloneable, Serializable and Object.
+     * Filter out the ones that do not have run-time type info at all, as
+     * we do for other classes.
+     */
+    val strictAncestors =
+      List(CloneableClass, SerializableClass, ObjectClass)
+        .filter(name => ctx.getClassInfoOption(name).exists(_.hasRuntimeTypeInfo))
+
+    val fb = newFunctionBuilder(genFunctionID.arrayTypeData)
+    val typeDataParam = fb.addParam("typeData", typeDataType)
+    val dimsParam = fb.addParam("dims", Int32)
+    fb.setResultType(objectVTableType)
+
+    val arrayTypeDataLocal = fb.addLocal("arrayTypeData", objectVTableType)
+
+    fb.loop() { loopLabel =>
+      fb.block(objectVTableType) { arrayOfIsNonNullLabel =>
+        // br_on_non_null $arrayOfIsNonNull typeData.arrayOf
+        fb += LocalGet(typeDataParam)
+        fb += StructGet(
+          genTypeID.typeData,
+          genFieldID.typeData.arrayOf
+        )
+        fb += BrOnNonNull(arrayOfIsNonNullLabel)
+
+        // <top-of-stack> := typeData ; for the <old typeData>.arrayOf := ... later on
+        fb += LocalGet(typeDataParam)
+
+        // typeData := new typeData(...)
+        fb += I32Const(0) // nameOffset
+        fb += I32Const(0) // nameSize
+        fb += I32Const(0) // nameStringIndex
+        fb += I32Const(KindArray) // kind = KindArray
+        fb += I32Const(0) // specialInstanceTypes = 0
+
+        // strictAncestors
+        for (strictAncestor <- strictAncestors)
+          fb += GlobalGet(genGlobalID.forVTable(strictAncestor))
+        fb += ArrayNewFixed(
+          genTypeID.typeDataArray,
+          strictAncestors.size
+        )
+
+        fb += LocalGet(typeDataParam) // componentType
+        fb += RefNull(HeapType.None) // name
+        fb += RefNull(HeapType.None) // classOf
+        fb += RefNull(HeapType.None) // arrayOf
+
+        // clone
+        fb.switch(RefType(genTypeID.cloneFunctionType)) { () =>
+          fb += LocalGet(typeDataParam)
+          fb += StructGet(genTypeID.typeData, genFieldID.typeData.kind)
+        }(
+          List(KindBoolean) -> { () =>
+            fb += ctx.refFuncWithDeclaration(genFunctionID.clone(BooleanRef))
```

Comment:
```suggestion
            fb += ctx.refFuncWithDeclaration(genFunctionID.cloneArray(BooleanRef))
```

It took me quite a while to figure out what was happening here.